librad/src/id/entity/data.rs

```diff
@@ -0,0 +1,209 @@
+use std::collections::{BTreeSet, HashMap, HashSet};
+
+#[derive(Clone, Deserialize, Serialize, Debug, PartialEq)]
+pub struct EntitySignatureData {
+    pub user: Option<String>,
+    pub sig: String,
+}
+
+fn ordered_set<S>(value: &HashSet<String>, serializer: S) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    let ordered: BTreeSet<String> = value.iter().cloned().collect();
+    ordered.serialize(serializer)
+}
+
+#[derive(Clone, Deserialize, Serialize, Debug, Default)]
+pub struct EntityData<T> {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub revision: Option<u64>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub hash: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub parent_hash: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub signatures: Option<HashMap<String, EntitySignatureData>>,
+
+    #[serde(
+        skip_serializing_if = "HashSet::is_empty",
+        serialize_with = "ordered_set",
+        default
+    )]
+    pub keys: HashSet<String>,
+    #[serde(
+        skip_serializing_if = "HashSet::is_empty",
+        serialize_with = "ordered_set",
+        default
+    )]
+    pub certifiers: HashSet<String>,
+
+    pub info: T,
+}
+
+impl<T> EntityData<T>
+where
+    T: Serialize + DeserializeOwned + Clone + Default,
+{
+    pub fn to_json_writer<W>(&self, writer: W) -> Result<(), Error>
+    where
+        W: std::io::Write,
+    {
+        serde_json::to_writer(writer, self).map_err(Error::SerializationFailed)?;
+        Ok(())
+    }
+    pub fn to_json_string(&self) -> Result<String, Error> {
+        Ok(serde_json::to_string(self).map_err(Error::SerializationFailed)?)
+    }
+
+    pub fn from_json_reader<R>(r: R) -> Result<Self, Error>
+    where
+        R: std::io::Read,
+    {
+        serde_json::from_reader(r).map_err(Error::SerializationFailed)
+    }
+    pub fn from_json_str(s: &str) -> Result<Self, Error> {
+        serde_json::from_str(s).map_err(Error::SerializationFailed)
+    }
+
+    pub fn canonical_data(&self) -> Result<Vec<u8>, Error> {
+        let mut cleaned = EntityData::<T>::default();
+        cleaned.name = self.name.to_owned();
+        cleaned.revision = self.revision.to_owned();
+        cleaned.hash = self.hash.to_owned();
+        cleaned.keys = self.keys.to_owned();
+        cleaned.certifiers = self.certifiers.to_owned();
+        cleaned.info = self.info.to_owned();
+
+        let mut buffer: Vec<u8> = vec![];
+        let mut ser =
+            serde_json::Serializer::with_formatter(&mut buffer, CanonicalFormatter::new());
+        cleaned
+            .serialize(&mut ser)
+            .map_err(Error::SerializationFailed)?;
+        Ok(buffer)
+    }
+
+    pub fn compute_hash(&self) -> Result<Multihash, Error> {
+        Ok(Sha2_256::digest(&self.canonical_data()?))
```

Comment:
Because it is unlikely to have collisions, and if I am not mistaken it is the next hash function that git will use (transitioning away from SHA1).

Reply:
But we don’t benefit in any way from git’s hashing, or do we? As SHA2 is already considered obsolete, I’d lean towards using a recent algorithm (either SHA3 or BLAKE3, which claims to be much faster). There’s also the consideration that hashing is likely to become a bottleneck for testing, so swapping for a cryptographically insecure but fast algorithm becomes desirable. As a follow-up, is this suggest to abstract over the hash algorithm, and introduce a “canonical” hashing module.

Reply:
+1, but I'd like to add the concern that it will likely be useful for the Registry to verify these documents at some point in the future, and there we are a lot more limited in terms of hashing functions (we have to compile them in the wasm blob), so using the same hashing function (Blake2b) would make things interop more easily.

Reply:
> Blake2b

Why is this, actually?

Reply:
It's the default in polkadot/substrate (it *might* be possible to change though), besides also being a really good hashing function.

Reply:
> it might be possible to change

Hm ok, good to know. Also dissapointed we can't use Blake2b++, being so much betterer

Reply:
I agree and I added one more checkbox in the #87 description to track this.